File: web-api/src/main/java/fi/vm/yti/groupmanagement/model/AuthenticatedUser.java
package fi.vm.yti.groupmanagement.model;

import fi.vm.yti.security.Role;
import fi.vm.yti.security.YtiUser;
import org.jetbrains.annotations.NotNull;

import java.util.*;

public class AuthenticatedUser {

    private final boolean anonymous;
    private final String email;
    private final String firstName;
    private final String lastName;
    private final boolean superuser;
    private final List<OrganizationRole> rolesInOrganizations = new ArrayList<>();

    public AuthenticatedUser(@NotNull YtiUser user) {

        this.anonymous = user.isAnonymous();
        this.email = user.getEmail();
        this.firstName = user.getFirstName();
        this.lastName = user.getLastName();
        this.superuser = user.isSuperuser();

        for (Map.Entry<UUID, Set<Role>> roleSetEntry : user.getRolesInOrganizations().entrySet()) {
            for (Role role : roleSetEntry.getValue()) {
                rolesInOrganizations.add(new OrganizationRole(role, roleSetEntry.getKey()));
            }
        }
    }

    public boolean isAnonymous() {
        return anonymous;
    }

    public String getEmail() {
        return email;
    }

    public String getFirstName() {
        return firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public boolean isSuperuser() {
        return superuser;
    }

    public List<OrganizationRole> getRolesInOrganizations() {
        return rolesInOrganizations;
    }

    public static class OrganizationRole {

        private final Role role;
        private final UUID organization;

        OrganizationRole(Role role, UUID organization) {
            this.role = role;
            this.organization = organization;
        }

        public Role getRole() {
            return role;
        }

        public UUID getOrganization() {
            return organization;
        }
    }
}

File: frontend/src/app/entities/authenticated-user.ts
import { AuthenticatedUser as ApinaAuthenticatedUser, UUID } from '../apina';

type Role = 'ADMIN'
          | 'DATA_MODEL_EDITOR'
          | 'TERMINOLOGY_EDITOR'
          | 'CODE_LIST_EDITOR';

export class AuthenticatedUser {

  anonymous: boolean;
  email: string;
  firstName: string;
  lastName: string;
  superuser: boolean;
  rolesInOrganizations = new Map<string, Set<Role>>();

  constructor(user: ApinaAuthenticatedUser) {
    this.anonymous = user.anonymous;
    this.email = user.email;
    this.firstName = user.firstName;
    this.lastName = user.lastName;
    this.superuser = user.superuser;

    for (const organizationRole of user.rolesInOrganizations) {
      this.getRolesInOrganization(organizationRole.organization).add(organizationRole.role as Role);
    }
  }

  private getRolesInOrganization(organizationId: UUID): Set<Role> {

    const roles = this.rolesInOrganizations.get(organizationId as string);

    if (roles) {
      return roles;
    } else {
      const newSet = new Set<Role>();
      this.rolesInOrganizations.set(organizationId as string, newSet);
      return newSet;
    }
  }

  isAdmin(organizationId: UUID): boolean {
    return this.getRolesInOrganization(organizationId).has('ADMIN');
  }

  isAdminInAnyOrganization() {

    for (const roles of Array.from(this.rolesInOrganizations.values())) {
      if (roles.has('ADMIN')) {
        return true;
      }
    }

    return false;
  }
}
